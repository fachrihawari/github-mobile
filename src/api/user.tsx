import axios from "axios";

import config from "../config";

export const profile = (username: string, password: string) => {
  return axios.get(`${config.apiUrl}/user`, {
    auth: {
      username,
      password
    }
  });
};
