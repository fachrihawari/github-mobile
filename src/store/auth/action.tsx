import { FETCH_USER_REQUEST, LOGOUT } from "./constant";
import { createAction } from "../helpers";

export const fetchUserRequest = (username: string, password: string, OTP: string='') =>
  createAction(FETCH_USER_REQUEST, { username, password, OTP });

export const logout = () => createAction(LOGOUT);
