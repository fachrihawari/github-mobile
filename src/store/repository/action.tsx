import { FETCH_COMMIT_REQUEST } from "./constant";
import { createAction } from "../helpers";

export const fetchCommitRequest = (
  repository: string,
  page?: number,
  perPage?: number
) => createAction(FETCH_COMMIT_REQUEST, { repository, page, perPage });
