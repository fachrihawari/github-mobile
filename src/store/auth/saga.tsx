import { AnyAction } from "redux";
import { takeLatest, put, call, delay } from "redux-saga/effects";

import {
  FETCH_USER_REQUEST,
  FETCH_USER_BEGIN,
  FETCH_USER_SUCCESS,
  FETCH_USER_FAILURE
} from "./constant";
import * as api from "../../api/user";
import { parseError } from "../helpers";

function* userRequest({ payload }: AnyAction) {
  try {
    yield put({ type: FETCH_USER_BEGIN });
    const { data } = yield call(
      api.profile,
      payload.username,
      payload.password,
      payload.OTP
    );
    yield put({
      type: FETCH_USER_SUCCESS,
      payload: {
        profile: data
      }
    });
  } catch (error) {
    console.log(error.response)
    
    let needOTP = false
    const { headers } = error.response
    if (headers['x-github-otp']) {
      const otpType = ['sms', 'app']
      const needType = headers['x-github-otp'].replace('required;', '').trim()
      needOTP = otpType.indexOf(needType) > -1

      try {
        yield call(
          api.authorize,
          payload.username,
          payload.password
        );
      } catch (error) {}

    }

    yield put({
      type: FETCH_USER_FAILURE,
      payload: {
        error: parseError(error),
        needOTP
      }
    });
  }
}
export default [takeLatest(FETCH_USER_REQUEST, userRequest)];
