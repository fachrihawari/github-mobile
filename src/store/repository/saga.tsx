import { AnyAction } from "redux";
import { takeLatest, put, call, delay } from "redux-saga/effects";

import {
  FETCH_COMMIT_REQUEST,
  FETCH_COMMIT_BEGIN,
  FETCH_COMMIT_SUCCESS,
  FETCH_COMMIT_FAILURE
} from "./constant";
import * as api from "../../api/repository";
import { parseError } from "../helpers";

function* commitRequest({ payload }: AnyAction) {
  try {
    yield put({ type: FETCH_COMMIT_BEGIN });
    const { data } = yield call(
      api.commits,
      payload.repository,
      payload.page,
      payload.perPage
    );
    yield put({
      type: FETCH_COMMIT_SUCCESS,
      payload: {
        commits: data,
        currentPage: payload.page < 1 ? 1 : payload.page
      }
    });
  } catch (error) {
    yield put({
      type: FETCH_COMMIT_FAILURE,
      payload: {
        error: parseError(error)
      }
    });
  }
}
export default [takeLatest(FETCH_COMMIT_REQUEST, commitRequest)];
