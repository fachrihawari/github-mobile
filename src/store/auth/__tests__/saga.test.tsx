import { SagaIterator } from 'redux-saga';
import { call, put } from 'redux-saga/effects';
import * as userApi from '../../../api/user';
import { fetchUserBegin, fetchUserFailure, fetchUserRequest, fetchUserSuccess } from '../action';
import { userRequest } from '../saga';

const profile = {
  'login': 'github',
  'id': 9919,
  'node_id': 'MDEyOk9yZ2FuaXphdGlvbjk5MTk=',
  'avatar_url': 'https://avatars1.githubusercontent.com/u/9919?v=4',
  'gravatar_id': '',
  'url': 'https://api.github.com/users/github',
  'html_url': 'https://github.com/github',
  'followers_url': 'https://api.github.com/users/github/followers',
  'following_url': 'https://api.github.com/users/github/following{/other_user}',
  'gists_url': 'https://api.github.com/users/github/gists{/gist_id}',
  'starred_url': 'https://api.github.com/users/github/starred{/owner}{/repo}',
  'subscriptions_url': 'https://api.github.com/users/github/subscriptions',
  'organizations_url': 'https://api.github.com/users/github/orgs',
  'repos_url': 'https://api.github.com/users/github/repos',
  'events_url': 'https://api.github.com/users/github/events{/privacy}',
  'received_events_url': 'https://api.github.com/users/github/received_events',
  'type': 'Organization',
  'site_admin': false,
  'name': 'GitHub',
  'company': null,
  'blog': 'https://github.com/about',
  'location': 'San Francisco, CA',
  'email': 'support@github.com',
  'hireable': null,
  'bio': 'How people build software.',
  'public_repos': 306,
  'public_gists': 0,
  'followers': 0,
  'following': 0,
  'created_at': '2008-05-11T04:37:31Z',
  'updated_at': '2019-07-15T14:26:15Z'
}
const credentials = {
  username: 'Username',
  password: 'Password',
  OTP: '',
}

describe('userRequest Saga', () => {
  let getUserGenerator: SagaIterator


  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {

    const action = fetchUserRequest(credentials.username, credentials.password, credentials.OTP)
    getUserGenerator = userRequest(action);
  });

  it('should dispatch the fetchUserSuccess action if it requests the data successfully', () => {

    const putBegin = getUserGenerator.next().value;
    expect(putBegin).toMatchSnapshot();
    expect(putBegin).toEqual(put(fetchUserBegin()));

    const callDescriptor = getUserGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
    expect(callDescriptor).toEqual(call(userApi.profile, credentials.username, credentials.password, credentials.OTP));
    
    const putFailure = getUserGenerator.next({ data: profile }).value;
    expect(putFailure).toMatchSnapshot();
    expect(putFailure).toEqual(put(fetchUserSuccess(profile)));
  });

  it('should dispatch the fetchUserFailure action', () => {
    const putBegin = getUserGenerator.next().value;
    expect(putBegin).toMatchSnapshot();
    expect(putBegin).toEqual(put(fetchUserBegin()));

    const callDescriptor = getUserGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
    expect(callDescriptor).toEqual(call(userApi.profile, credentials.username, credentials.password, credentials.OTP));

    const error = {
      response: {
        data: {
          message: 'Bad Request'
        },
        headers: {}
      }
    }
    // @ts-ignore
    const putFailure = getUserGenerator.throw(error).value 
    expect(putFailure).toMatchSnapshot();
    expect(putFailure).toEqual(put(fetchUserFailure('Bad Request', false)));
  });

  it('should dispatch the fetchUserFailure action if 2FA enabled', () => {
    const putBegin = getUserGenerator.next().value;
    expect(putBegin).toMatchSnapshot();
    expect(putBegin).toEqual(put(fetchUserBegin()));

    const profileCallDescriptor = getUserGenerator.next().value;
    expect(profileCallDescriptor).toMatchSnapshot();
    expect(profileCallDescriptor).toEqual(call(userApi.profile, credentials.username, credentials.password, credentials.OTP));

    const error = {
      response: {
        data: {
          message: 'Must specify two-factor authentication OTP code.',
          documentation_url: 'https://developer.github.com/v3/auth#working-with-two-factor-authentication'
        },
        headers: {
          'x-github-otp': 'required; sms'
        }
      }
    }
    
    // @ts-ignore
    const authorizeCallDescriptor = getUserGenerator.throw(error).value;
    expect(authorizeCallDescriptor).toMatchSnapshot();
    expect(authorizeCallDescriptor).toEqual(call(userApi.authorize, credentials.username, credentials.password));
    
    const putFailure = getUserGenerator.next(error).value;    
    expect(putFailure).toMatchSnapshot();
    expect(putFailure).toEqual(put(fetchUserFailure('Must specify two-factor authentication OTP code.', true)));
    
  });
  
  it('should dispatch the fetchUserFailure action if failing network', () => {
    const putBegin = getUserGenerator.next().value;
    expect(putBegin).toMatchSnapshot();
    expect(putBegin).toEqual(put(fetchUserBegin()));

    const callDescriptor = getUserGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
    expect(callDescriptor).toEqual(call(userApi.profile, credentials.username, credentials.password, credentials.OTP));
    
    const putFailure = getUserGenerator.next().value;
    expect(putFailure).toMatchSnapshot();
    expect(putFailure).toEqual(put(fetchUserFailure('You\'re offline!', false)));
  });
});