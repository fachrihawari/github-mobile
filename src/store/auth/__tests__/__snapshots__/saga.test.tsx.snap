// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userRequest Saga should dispatch the fetchUserFailure action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "auth/FETCH_USER_BEGIN",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`userRequest Saga should dispatch the fetchUserFailure action 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "Username",
      "Password",
      "",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`userRequest Saga should dispatch the fetchUserFailure action 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": Object {
        "error": "Bad Request",
        "needOTP": false,
      },
      "type": "auth/FETCH_USER_FAILURE",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`userRequest Saga should dispatch the fetchUserFailure action if 2FA enabled 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "auth/FETCH_USER_BEGIN",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`userRequest Saga should dispatch the fetchUserFailure action if 2FA enabled 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "Username",
      "Password",
      "",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`userRequest Saga should dispatch the fetchUserFailure action if 2FA enabled 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "Username",
      "Password",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`userRequest Saga should dispatch the fetchUserFailure action if 2FA enabled 4`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": Object {
        "error": "Must specify two-factor authentication OTP code.",
        "needOTP": true,
      },
      "type": "auth/FETCH_USER_FAILURE",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`userRequest Saga should dispatch the fetchUserFailure action if failing network 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "auth/FETCH_USER_BEGIN",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`userRequest Saga should dispatch the fetchUserFailure action if failing network 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "Username",
      "Password",
      "",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`userRequest Saga should dispatch the fetchUserFailure action if failing network 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": Object {
        "error": "You're offline!",
        "needOTP": false,
      },
      "type": "auth/FETCH_USER_FAILURE",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`userRequest Saga should dispatch the fetchUserSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "auth/FETCH_USER_BEGIN",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`userRequest Saga should dispatch the fetchUserSuccess action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "Username",
      "Password",
      "",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`userRequest Saga should dispatch the fetchUserSuccess action if it requests the data successfully 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": Object {
        "profile": Object {
          "avatar_url": "https://avatars1.githubusercontent.com/u/9919?v=4",
          "bio": "How people build software.",
          "blog": "https://github.com/about",
          "company": null,
          "created_at": "2008-05-11T04:37:31Z",
          "email": "support@github.com",
          "events_url": "https://api.github.com/users/github/events{/privacy}",
          "followers": 0,
          "followers_url": "https://api.github.com/users/github/followers",
          "following": 0,
          "following_url": "https://api.github.com/users/github/following{/other_user}",
          "gists_url": "https://api.github.com/users/github/gists{/gist_id}",
          "gravatar_id": "",
          "hireable": null,
          "html_url": "https://github.com/github",
          "id": 9919,
          "location": "San Francisco, CA",
          "login": "github",
          "name": "GitHub",
          "node_id": "MDEyOk9yZ2FuaXphdGlvbjk5MTk=",
          "organizations_url": "https://api.github.com/users/github/orgs",
          "public_gists": 0,
          "public_repos": 306,
          "received_events_url": "https://api.github.com/users/github/received_events",
          "repos_url": "https://api.github.com/users/github/repos",
          "site_admin": false,
          "starred_url": "https://api.github.com/users/github/starred{/owner}{/repo}",
          "subscriptions_url": "https://api.github.com/users/github/subscriptions",
          "type": "Organization",
          "updated_at": "2019-07-15T14:26:15Z",
          "url": "https://api.github.com/users/github",
        },
      },
      "type": "auth/FETCH_USER_SUCCESS",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;
